dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java

```diff
@@ -832,29 +866,42 @@ boolean handleOutliers(ThroughputDistribution distribution, double throughput) {
             Integer priorKey = Integer.valueOf(poolSize);
             int smallerPools = 0;
             while (true) {
-                int distance = 0;
                 // stop if we run out of data
                 if (shrinkKey == null)
                     break;
                 priorKey = shrinkKey;
                 shrinkKey = threadStats.lowerKey(shrinkKey);
+                int distance = 0;
+                boolean pruned = false;
+                boolean inHangBuffer = true;
                 // discard invalid data (old or out-of-range) found in comparison stats
                 ThroughputDistribution priorStats = threadStats.get(priorKey);
                 distance = poolSize - priorKey.intValue();
                 if (pruneData(priorStats, forecast, distance, downwardCompareSpan)) {
-                    threadStats.remove(priorKey);
-                } else {
+                    if (priorKey > hangBufferPoolSize) {
+                        threadStats.remove(priorKey);
+                        pruned = true;
+                        inHangBuffer = false;
+                    }
+                }
+                if (!pruned) {
                     // found a valid smaller poolSize, so include that data in the shrinkScore
                     smallerPools++;
                     shrinkScore += priorStats.getProbabilityGreaterThan(forecast);
-                    // stop if reached/passed compareSpan
-                    if (distance >= downwardCompareSpan)
-                        break;
+                    if (inHangBuffer) {
+                        // stop when we get compareRange datapoints
+                        if (smallerPools >= compareRange)
+                            break;
+                    } else {
+                        // stop if reached/passed compareSpan
+                        if (distance >= downwardCompareSpan)
+                            break;
+                    }
                 }
             }
 
             // if we didn't find compareRange datapoints, flip a coin
-            if (smallerPools < compareRange) {
+            if ((smallerPools < compareRange) && (priorKey > coreThreads)) {
```

Comment:
Do we need to compare against the hangBufferPoolSize?

Reply:
No ... though this is a bit subtle also ... we want to allow the controller to move the poolSize below the hangBufferPoolSize if the data (or lack thereof, in this case) suggests that is a better place to be. If we never allow the poolSize to go below hangBufferPoolSize, then we may as well just increase coreThreads when a hang is detected ... that seems to be undesirably rigid, since a transient hang issue might leave us stuck at a larger-than-optimal poolsize.
So in other areas of the controller we avoid extra influences toward shrinking, based e.g. on cpuHigh, but in the basic shrinkScore calculation we allow the possibility of shrinking below hangBufferPoolSize ... it is a bit of a balancing act, not sure if this will be the final/best solution but seems to work well so far.